README.md
# joj-scoreboard-hacker
A tool to get scoreboard for a homework on [joj](https://joj.sjtu.edu.cn/) even if TA closed the scoreboard.



## prerequisites

python 3, together with python library

- requests
- pandas

install them with command

``` bash
pip3 install requests
pip3 install pandas
```



## how to use

Run the `main.py`, and then enter the prompt message

```bash
wonderwind@DESKTOP-53BM3RL:/mnt/c/Users/X1G8/Desktop/newrepo/joj-scoreboard-hacker$ python3 main.py
please input the main page url of the course, for example, https://joj.sjtu.edu.cn/d/ve280_summer_2022_paul/: https://joj.sjtu.edu.cn/d/ve280_summer_2022_paul/
please input the homework id (not problem id) of the course, for example, 62d078c1dec511000876c05d: 62d078c1dec511000876c05d
please input the problem id of the course, for example, 62c77752dec511000876987b:
```

enter the main page url, the hw id and the problem id. Main page url is the url of this course when you're at the main page. Homework id id is the id after the main page url when you click into a homework. For example, we can see the url of VE280 SU22 project 5 is https://joj.sjtu.edu.cn/d/ve280_summer_2022_paul/homework/62d078c1dec511000876c05d, so the homework id is "62d078c1dec511000876c05d".

Problem id is useful when a homework contains multiple problems. However, you can enter an empty line when the homework only contains one problem. Just as an example, the problem url of the VE280 SU22 project 5 is https://joj.sjtu.edu.cn/d/ve280_summer_2022_paul/homework/62d078c1dec511000876c05d/62c77752dec511000876987b, so the only problem's problem id is "62c77752dec511000876987b". You can of course, leave an empty message here.

Next, the program will ask you to enter the database of students.

```bash
please input the selection of students, in 18, 19, 20, 21 entry, for example, 0011 means choose 20 and 21 entry: 1110
```

Here, the input format is always $a_1a_2a_3a_4$, where $a_i=0 \text{ or } 1$ for $i=1,2,3,4$. When $a_i=1$, the corresponding grade (年级) of student is selected. For example, when we enter "1110" here, the student in 2018, 2019, 2020 entry are selected.

The program will further ask you if you want to also use previous data for this course.

```
please input if you want to also use previous data: 1
```

The program always keep a name list of all the students that appears in this course at least once from previous run of this course. If you choose "1" here, the database we choose is the **union** of what you choose (selection of students previously by "1110") and all the students that appear in the previous run. 



The program will then start to give students' score in this homework.

```
student 518370910034 金彤: total case: 85, accepted case: 85, score: 100.0
student 518370910107 ***: total case: 85, accepted case: 47, score: 55.294117647058826
student 518021910448 ***: total case: 85, accepted case: 16, score: 18.823529411764707
student 518021911016 ***: total case: 85, accepted case: 15, score: 17.647058823529413
student 518370990038 ***: total case: 85, accepted case: 85, score: 100.0
student 519370910050 ***: total case: 85, accepted case: 57, score: 67.05882352941177
...
```

The "***" here is the students' name. 

Note that the speed in one grade of student is determined by **the total number of students in that grade, not the total number that enrolled in this course**. Since very few students in 2018 entry choose VE280 this semester, we can see that the speed when one line of student's score is given at a lower speed than in 2020 entry.

Finally, when all the students' score are shown, you'll be prompt to export the data to a csv file.

```
student 520370910134 **: total case: 85, accepted case: 85, score: 100.0
student 520370910135 ***: total case: 85, accepted case: 46, score: 54.11764705882353
student 520370910142 ***: total case: 85, accepted case: 85, score: 100.0
please enter the csv name. Empty means abort. No need to add .csv: VE280_SU22_p5
data exported to VE280_SU22_p5.csv
```

Here, we can enter the name of the exported file. Here we enter "VE280_SU22_p5", and we get a csv file "VE280_SU22_p5.csv" in the current folder with all the data above.

If you do not want the csv, just enter an empty line, and the csv will not be generated. 



## About the namelist function

As is described above, the program will always keep a namelist of all the students that appears in this course at least once. This function can greatly reduce the amount of time, because if we run full test, the program will send request to joj server for all the students, not only for the students in the course.

By entering "0000" when selecting the database, and enter "1" to use the namelist, you can run all the students that appeared before. This will save a lot of time and miss at most 2-5 students.

For a typical "5 project" course (like VE280 SU22 and FA21), I suggest you run full test (enter "1110" for database selection) for the p1 and p2. Then just use the namelist (enter "0000" for the database and "1" for using previous data). That's because it's unlikely that a student miss the first 2 project (relatively easy) but do the projects later.

For the course that involves using joj for exam, you can run full test for the first problem of the exam, and always just use the namelist for all the rest assignments and exams. That's because it's likely that everyone take part in the exam.



## further notice

The speed is subject to your internet speed and joj server's respond speed. It's not related to your CPU.

The principle of this program is to take advantage of joj's "recent submission" function, and try each student in the database if you do not use namelist function.

Do not abuse it since it may burden the joj server.

Also, it seems that the joj will be upgraded to 2.0. I'm not sure if this program will still work at that time.


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.1.0] alpha release: Aug 11, 2022

Basic functions are completed.

- Generate the scoreboard
- export as csv
- choose database of different grades. 

## [0.2.0] beta release: Aug 12, 2022

Can record namelist of a course, 
and reuse them in later runs.

Set `./local` as the folder to store all the histories. `./local/[course_url]`
folder is used to store the materials in one course

demo.py
# a demo to show the working principle of joj scoreboard hacker

import requests

# first prompt to get information. stud_id can be set within a range
stud_id = 520370910097  # I use my own sjtu id to avoid privacy leak here
base_url = "https://joj.sjtu.edu.cn/d/ve280_summer_2022_paul/"  # VE280 SU22
assignment = "62d078c1dec511000876c05d"

# start search process
url = base_url + "records?uid_or_name=" + str(stud_id) + "&pid=&tid=" + assignment
r = requests.get(url)  # can use param, but let's be straightforward here
res1 = str(r.content)
place1 = res1.find("data-rid") # place==-1 means id not found. will be tackled later
place_start = res1.find('\"', place1)
place_end = res1.find('\"', place_start+1)
link_id = res1[place_start+1:place_end]
url_score_page = base_url + "records/" + link_id
print(url_score_page)  # link to the first record of the student in this assignment. It's the sentence test
r2 = requests.get(url_score_page)
res2 = str(r2.content)

# count ac and all results. if "Compile Error" is found, get 0 directly
AC_count = res2.count("icon pass")
count = res2.count("icon pass") + res2.count("icon fail")
if AC_count == count:
    AC_count -= 1
    count -= 1
else:
    count -= 1
print(f"total case: {count}, accepted case: {AC_count}, score: {100*AC_count/count}")



local_file.py
import os


def mkdir(path):
    if not os.path.exists(path):
        os.makedirs(path)


def fit_url(url: str):
    lurl = list(url)
    for i in range(len(lurl)):
        if lurl[i] == '/' or lurl[i] == ':':
            lurl[i] = '_'
    return "".join(lurl)


# store namelist at local/base_url(: and / become _)
def id_store(all_stud_id, base_url):
    folder = "local/" + fit_url(base_url)
    mkdir(folder)
    namelist = set()
    file_name = folder + "/namelist.txt"
    if os.path.isfile(file_name):  # load data
        fp = open(file_name, "r")
        names = fp.read().split("\n")
        if len(names):
            if (names[-1] == ""):
                names.pop()
        namelist = set(names)
        fp.close()
    namelist = namelist.union(map(str, all_stud_id))
    fp = open(file_name, "w")
    for line in namelist:
        fp.write(line+"\n")
    fp.close()


def id_load(base_url):
    folder = "local/" + fit_url(base_url)
    file_name = folder + "/namelist.txt"
    if os.path.isfile(file_name):
        fp = open(file_name, 'r')
        names = fp.read().split("\n")
        if len(names):
            if (names[-1] == ""):
                names.pop()
        return names
    else:
        return []


unit_quire.py
# -*- coding: UTF-8 -*-
import requests


def quire(stud_id, base_url, assignment, problem):
    '''
    :param stud_id: student's sjtu id
    :param base_url: the main page url of this course
    :param assignment: the id of the homework (not sub-problem)
    :param problem: the id of the sub-problem
    :return: a tuple (case_num, AC_count, score, state, name), state 1: Compile error, 2: not found, 3: get score
    '''
    # start search process
    url = base_url + "records?uid_or_name=" + str(stud_id) + "&pid=" + problem + "&tid=" + assignment
    r = requests.get(url)  # can use param, but let's be straightforward here
    res1 = str(r.content)
    place1 = res1.find("data-rid")  # place==-1 means id not found. will be tackled later
    place_start = res1.find('\"', place1)
    place_end = res1.find('\"', place_start + 1)
    link_id = res1[place_start + 1:place_end]
    url_score_page = base_url + "records/" + link_id
    # print(url_score_page)  # link to the first record of the student in this assignment. It's the sentence test
    r2 = requests.get(url_score_page)
    res2 = str(r2.content, 'utf8')
    # get username
    username = ""
    place_user_label = res2.find("user-profile-name")
    if place_user_label >= 0:
        l1 = res2.find(">", place_user_label)
        l2 = res2.find("<", l1)
        username = res2[l1 + 1:l2]
    # count ac and all results. if "Compile Error" is found, get 0 directly
    if (res2.count("Compile Error")):
        # print("compile error")
        return (0, 0, 0, 1, username)
    AC_count = res2.count("icon pass")
    count = res2.count("icon pass") + res2.count("icon fail")
    if AC_count == count:
        AC_count -= 1
        count -= 1
    else:
        count -= 1
    if count <= 0:  # student not found
        return (-1, -1, -1, 2, username)
    score = 100 * AC_count / count
    # print(f"total case: {count}, accepted case: {AC_count}, score: {score}")
    return (count, AC_count, score, 3, username)


main.py
# -*- coding: UTF-8 -*-
from unit_quire import *
from local_file import *
import pandas as pd

base_url = input("please input the main page url of the course, for example, "
                 "https://joj.sjtu.edu.cn/d/ve280_summer_2022_paul/: ")
assignment = input("please input the homework id (not problem id) of the course, for example, "
                   "62d078c1dec511000876c05d: ")
problem = input("please input the problem id of the course, for example, "
                "62c77752dec511000876987b: ")
id_2020 = open("2020_id.txt", "r")
id_2021 = open("2021_id.txt", "r")
id_2019 = open("2019_id.txt", "r")
id_2018 = open("2018_id.txt", "r")
sel = input("please input the selection of students, in 18, 19, 20, 21 entry, for example, "
            "0011 means choose 20 and 21 entry: ")
all_id = []
if sel[0] == '1':
    all_id += id_2018.read().split('\n')
if sel[1] == '1':
    all_id += id_2019.read().split('\n')
if sel[2] == '1':
    all_id += id_2020.read().split('\n')
if sel[3] == '1':
    all_id += id_2021.read().split('\n')
id_2020.close()
id_2021.close()
id_2019.close()
id_2018.close()
sel = input("please input if you want to also use previous data: ")
if sel:
    all_id += id_load(base_url)
    all_id = list(set(all_id))

all_id.sort()
data = []
save_id = []
for stud_id in all_id:
    (case_num, AC_count, score, state, username) = quire(stud_id, base_url, assignment, problem)
    if state == 3:
        print(f"student {stud_id} {username}: total case: {case_num}, accepted case: {AC_count}, score: {score}")
        data.append([stud_id, username, case_num, AC_count, score])
        save_id.append(stud_id)
    elif state == 1:
        print(f"student {stud_id} {username}: compile error")
        data.append([stud_id, username, "compile error", 0, 0])
        save_id.append(stud_id)
id_store(save_id, base_url)

df = pd.DataFrame(data, index=range(1, len(data) + 1), columns=["sjtu id", "name", "total case", "AC case", "score"])
output_filename = input("please enter the csv name. Empty means abort. No need to add .csv: ")
if output_filename != "":
    df.to_csv("local/" + fit_url(base_url)+"/"+output_filename+".csv")
    print("data exported to "+output_filename+".csv")
else:
    print("data not exported")
